lib/Sema/CSApply.cpp

```diff
@@ -9243,44 +9254,118 @@ static llvm::Optional<SyntacticElementTarget> applySolutionToForEachStmt(
     stmt->setWhere(rewrittenTarget->getAsExpr());
   }
 
-  // Convert that llvm::Optional<Element> value to the type of the pattern.
-  auto optPatternType = OptionalType::get(forEachStmtInfo.initType);
-  Type nextResultType = OptionalType::get(forEachStmtInfo.elementType);
-  if (!optPatternType->isEqual(nextResultType)) {
-    ASTContext &ctx = cs.getASTContext();
-    OpaqueValueExpr *elementExpr = new (ctx) OpaqueValueExpr(
-        stmt->getInLoc(), nextResultType->getOptionalObjectType(),
-        /*isPlaceholder=*/true);
-    Expr *convertElementExpr = elementExpr;
-    if (TypeChecker::typeCheckExpression(
-            convertElementExpr, dc,
-            /*contextualInfo=*/{forEachStmtInfo.initType, CTP_CoerceOperand})
-            .isNull()) {
+  return info;
+}
+
+static llvm::Optional<PackIterationInfo> applySolutionToForEachStmt(
+    Solution &solution, ForEachStmt *stmt, PackIterationInfo info,
+    llvm::function_ref<
+        llvm::Optional<SyntacticElementTarget>(SyntacticElementTarget)>
+        rewriteTarget) {
+
+  // A special walker to record opened element environment for var decls in a
+  // for-each loop.
+  class Walker : public ASTWalker {
+    GenericEnvironment *Environment;
+
+  public:
+    Walker(GenericEnvironment *Environment) { this->Environment = Environment; }
+
+    PreWalkResult<Stmt *> walkToStmtPre(Stmt *S) override {
+      if (isa<ForEachStmt>(S)) {
+        return Action::Stop();
+      }
+      return Action::Continue(S);
+    }
+
+    PreWalkAction walkToDeclPre(Decl *D) override {
+      if (auto *decl = dyn_cast<VarDecl>(D)) {
+        decl->setOpenedElementEnvironment(Environment);
```

Comment:
Thank you for the feedback! This change was necessary to enable var declarations inside the `for`-`repeat` loop, basically whenever you would declare (not reference) a variable inside of it, the AST Verifier crashed because of the `VarDecl::getTypeInContext` returning the type of the outer context. Also, the local variable binding in the source of the loop gets that environment as well.
Since a `ForEachStmt` is not a `DeclContext`, it, therefore, cannot have any generic context associated with it (that situation would be ideal, however, conforming to `DeclContext` will be a big change and people I asked about this seemed hesitant about this idea). 

In the example you provided, why do you think `x` will get the unneeded environment? The `Walker` sets the environment only if it encounters `VarDecl` *inside* the loop. And since the `x` inside the loop is a `DeclRefExpr`, it shouldn't get the unneeded environment, and therefore `VarDecl::getTypeInContext` will treat it normally.

During constraint generation, I thought we are not guaranteed to have an environment. 

In your example, I also am not sure if we actually need to set the environment for the references inside the loop? With the current implementation, I can already assign the elements of the pack being iterated over to a variable that was declared outside of the loop, and it works fine.

If you think I should not have this `Walker`, it'll be great if you provide a more detailed suggestion on how to replace it🥲

Reply:
Hm, maybe I looked at the outdated version but I thought it was updating declarations it fetched from `DeclRefExpr`s that's why I left this comment...

Using `walkToDeclPre` is better indeed going to handle only things which are declared in the loop but you need to be very careful about actually covering all of the declarations - we have multiple flag methods prefixed with`should*` associated with the `ASTWalker` and if you don't set all the flags correctly the walker might miss some of the declarations inside of other statements and expressions (i.e. string interpolations declare `$interpolation` variable implicitly and walking into them wasn't always enabled by default).

Reply:
Okay, I'll look into this, thanks🙂